hail/python/hail/matrixtable.py

```diff
@@ -539,6 +539,138 @@ class MatrixTable(ExprContainer):
+        +---------+-------------+-------------+
+        |   int32 |       int32 |       int32 |
+        +---------+-------------+-------------+
+        |       0 |           1 |           2 |
+        |       1 |           3 |           4 |
+        +---------+-------------+-------------+
+
+        Parameters
+        ----------
+        globals : :class:`dict` from :class:`str` to :obj:`any`
+            Global fields by name
+
+        rows: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Row fields by name
+
+        cols: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Column fields by name
+
+        entries: :class:`dict` from :class:`str` to :class:`list` of :class:`list` of :obj:`any`
+            Matrix entries by name in the form `entry[row_idx][col_idx]`
+
+        Returns
+        -------
+        :class:`.MatrixTable`
+            A MatrixTable assembled from inputs keyed by `row_idx`
+
+        Notes
+        -----
+        - The number of rows and columns specified in `rows` and `cols` for all
+          fields must match the matrix dimensions in `entries`.
+        - This method is intended for educational and testing purposes only.
+        """
+        # General idea: build a `Table` representation matching that returned by
+        # `MatrixTable.localize_entries` and then call `_unlocalize_entries`.
+        def invert(kvs: Dict[str, Iterable[Any]]) -> List[Dict[str, Any]]:
+            return [dict(zip(kvs, vs)) for vs in zip(*kvs.values())]
+
+        def anyval(kvs):
+            return next(iter(kvs.values()))
+
+        # In the case rows or cols aren't specified, we need to infer the
+        # matrix dimensions from *an* entry. Which one isn't important as we
+        # won't enforce congruence among input dimensions.
+        assert not ((rows is None or cols is None) and (entries is None))
+        cols = invert(cols) if cols else [{} for _ in anyval(entries)[0]]
+        for i, _ in enumerate(cols):
+            cols[i] = hl.struct(col_idx=i, **cols[i])
+
+        # While globals aren't required, `MatrixTable._unlocalize_entries`
+        # extracts matrix columns from the the `Table`'s globals.
+        if globals is None:
+            globals = {}
+
+        cols_field_name = 'columns'
+        globals[cols_field_name] = cols
+
+        # `MatrixTable._unlocalize_entries` extracts matrix entries from the
+        # rows on the table, so these need to be defined too.
+        rows = invert(rows) if rows else [{} for _ in anyval(entries)]
+        entries = map(invert, invert(entries)) if entries else [[{} for _ in cols] for _ in rows]
+
+        entries_field_name = 'entry_structs'
+        for i, (row_pros, entry_props) in enumerate(zip(rows, entries)):
+            row_entries = [hl.struct(**kvs) for kvs in entry_props]
+            rows[i] = hl.struct(row_idx=i, **row_pros, **{entries_field_name: row_entries})
```

Comment:
This is a bit of a technical issue, but I'm worried that this generates O(n_rows * n_cols) IR nodes. For example,
```
from_parts(rows=dict(x=[1,2],y=[3,4]), entries=dict(z=[[1], [2]]))
```
Will generate an IR like this:
```
(CastTableToMatrix
    ...
    (TableParallelize None
        (MakeStruct
            (rows 
                (MakeArray Array[Struct{row_idx:Int32,x:Int32,y:Int32,entry_structs:Array[Struct{z:Int32}]}] 
                    (MakeStruct 
                        (row_idx  (I32 0)) (x  (I32 1)) (y  (I32 3))
                        (entry_structs  (MakeArray Array[Struct{z:Int32}]  (MakeStruct  (z  (I32 1))))))
                    (MakeStruct
                        (row_idx  (I32 1)) (x  (I32 2)) (y  (I32 4))
                        (entry_structs  (MakeArray Array[Struct{z:Int32}]  (MakeStruct  (z  (I32 2))))))))
    ...)
```
If you instead use `hl.Struct` (which is just a Python object), Hail will encode this as a `Literal`, which should provide faster serialization, less compiler overhead, and, I think, faster runtime code (because the generated JVM bytecode doesn't reify our data as control flow). For example, just changing `hl.struct` to `hl.Struct` in this function has this effect:
```
...
(TableParallelize None
    (MakeStruct
        (rows
            (Literal Array[Struct{row_idx:Int32,x:Int32,y:Int32,entry_structs:Array[Struct{z:Int32}]}]
                "[{\\"row_idx\\": 0, \\"x\\": 1, \\"y\\": 3, \\"entry_structs\\": [{\\"z\\": 1}]}, {\\"row_idx\\": 1, \\"x\\": 2, \\"y\\": 4, \\"entry_structs\\": [{\\"z\\": 2}]}]"))
...
```
I'm a bit surprised to see this rendered as a JSON string. I thought these went through a fast binary format.

Not required for this pull request, but we could push this approach even further to create, I think, substantially faster code by instead serializing arrays as literals which we just index into with the row_idx and/or col_idx.

Reply:
@tpoterba am I making a mountain of an anthill about this?

Reply:
We'll punt on this, it can always be a follow up PR.